cmd/calico-ipam-upgrade/main.go

```diff
@@ -0,0 +1,231 @@
+
+	// Disable cni
+	if clusterInfo, err = calicoClient.ClusterInformation().Get(ctxt, "default", options.GetOptions{}); err != nil {
+		msg = "failed to fetch cluster information"
+		return
+	}
+	if *clusterInfo.Spec.DatastoreReady {
+		*clusterInfo.Spec.DatastoreReady = false
+		if clusterInfo, err = calicoClient.ClusterInformation().Update(ctxt, clusterInfo, options.SetOptions{}); err != nil {
+			msg = "failed to disable cluster"
+			return
+		}
+	}
+
+	// Establishing a mapping of IP addresses to Pods on this node
+	if podList, err = k8sClient.CoreV1().Pods("").List(v1.ListOptions{
+		LabelSelector: fmt.Sprintf("spec.nodeName = %s", nodename),
+	}); err != nil {
+		msg = "failed to list pods"
+		return
+	}
+	podIPMap := make(map[string]*corev1.Pod)
+	for i := 0; i < len(podList.Items); i++ {
+		podIPMap[podList.Items[i].Status.PodIP] = &podList.Items[i]
+	}
+
+	// Migrate IPIP tunnel address.
+	if node, err = calicoClient.Nodes().Get(ctxt, nodename, options.GetOptions{}); err != nil {
+		msg = "failed to get node resource"
+		return
+	}
+	ipipTunnelAddr := cnet.ParseIP(node.Spec.BGP.IPv4IPIPTunnelAddr)
+	if err = calicoClient.IPAM().AssignIP(ctxt, ipam.AssignIPArgs{
+		IP:       *ipipTunnelAddr,
+		Hostname: nodename,
+	}); err != nil {
+		msg = "failed to get add IPIP tunnel addr " + node.Spec.BGP.IPv4IPIPTunnelAddr
+		return
+	}
+
+	// Read in all the files in the host-local directory.
+	if files, err = ioutil.ReadDir(ipAllocPath); err != nil {
+		msg = "failed to read path " + ipAllocPath
+		return
+	}
+
+	// For each file, convert it into an IP allocation and then delete the file.
+	for _, f := range files {
+		fname := path.Join(ipAllocPath, f.Name())
+
+		// Delete and skip the last reserved IP.
+		if strings.Contains(f.Name(), "last") {
+			if err = os.Remove(fname); err != nil {
+				msg = "failed to remove file " + fname
+				return
+			}
+			continue
+		}
+
+		// The name of the file is its IP address.
+		if ip, _, err = cnet.ParseCIDR(fmt.Sprintf("%s/32", f.Name())); err != nil {
+			msg = "failed to parse IP " + f.Name()
+			return
+		}
+
+		// The contents are the container ID.
+		if b, err = ioutil.ReadFile(fname); err != nil {
+			msg = "failed to read file " + fname
+			return
+		}
+
+		// Get the pod resource associated with the IP.
+		pod, ok := podIPMap[f.Name()]
+		if !ok {
+			log.WithField("ip", f.Name()).Info("pod not found for IP, deleting and continuing...")
+			if err = os.Remove(fname); err != nil {
+				msg = "failed to remove file " + fname
+				return
+			}
+			continue
+		}
+
+		// Store allocation to datastore
+		handleID := fmt.Sprintf("%s.%s", "k8s-pod-network", string(b))
+		if err = calicoClient.IPAM().AssignIP(ctxt, ipam.AssignIPArgs{
+			IP:       *ip,
+			HandleID: &handleID,
+			Hostname: nodename,
+			Attrs: map[string]string{
+				"node":      nodename,
+				"pod":       pod.Name,
+				"namespace": pod.Namespace,
+			},
+		}); err != nil {
+			msg = "failed to assign IP to calico backend"
```

Comment:
What happens if we succeed here, but then fail to delete the file?

I think we'll retry, but when we get to this point a second time it will error since the IP is already allocated. We should probably special case the "Already exists" error case.